java/runtime/src/main/java/org/ray/runtime/object/ObjectSerializer.java

```diff
@@ -72,12 +81,16 @@ public static NativeRayObject serialize(Object object) {
     } else if (object instanceof byte[]) {
       // If the object is a byte array, skip serializing it and use a special metadata to
       // indicate it's raw binary. So that this object can also be read by Python.
-      return new NativeRayObject((byte[]) object, RAW_TYPE_META);
+      return new NativeRayObject((byte[]) object, OBJECT_METADATA_TYPE_RAW);
     } else if (object instanceof RayTaskException) {
-      return new NativeRayObject(Serializer.encode(object),
-          TASK_EXECUTION_EXCEPTION_META);
+      Serializer.Meta meta = new Serializer.Meta();
+      byte[] serializedBytes = Serializer.encode(object, meta);
+      return new NativeRayObject(serializedBytes, TASK_EXECUTION_EXCEPTION_META);
     } else {
-      return new NativeRayObject(Serializer.encode(object), null);
+      Serializer.Meta meta = new Serializer.Meta();
+      byte[] serializedBytes = Serializer.encode(object, meta);
+      return new NativeRayObject(serializedBytes, meta.isCrossLanguage ?
+          OBJECT_METADATA_TYPE_CROSS_LANGUAGE : OBJECT_METADATA_TYPE_JAVA);
```

Comment:
> It looks strange to me that `meta` is modified in `Serializer.encode`. Please return the metadata information in the return value.

I will return a pair contains serialized bytes and meta info as @raulchen suggests. But, in this way, deserializer can't takes serializer's output as input.

Reply:
> In my opinion, it's better to separate the `Serializer` class into three classes: `ExceptionSerializer`, `MessagePackSerializer`, `FstSerializer`.
> 
> Treat `ObjectSerializer` as a dispatcher. When serializing, it dispatches the serialization request into different serializer classes based on the object type and other information. When deserializing, it dispatches the deserialization request into different serializer classes based on the metadata. The dispatching logic should be entirely in `ObjectSerializer` so that it's more readable and maintainable.
> 
> The three different serializer classes may share some common logic, but it should be hidden from the `ObjectSerializer`.

The Fst serialized data  is embedded in MessagePack extension type. We can't simply dispatch the serialization and deserialization request before MessagePack's serialization / deserialization.